src/include/miopen/solver/implicitgemm_ck_util.hpp

```diff
@@ -165,5 +172,584 @@ ConvSolution InitAnyInvokerFactory(const ProblemDescriptionType& problem,
+            }
+            else
+            {
+                return SolverType{ctx,
+                                  problem.GetOutDataType(),
+                                  static_cast<uint32_t>(ck_args.N),
+                                  static_cast<uint32_t>(ck_args.K1),
+                                  static_cast<uint32_t>(ck_args.Ho),
+                                  static_cast<uint32_t>(ck_args.Wo)};
+            }
+        }
+    }
+};
+
+// Shorthand aliases for CK assuming CK always expects and generates NHWC/NDHWC layouts
+template <int ND, ConvOperandTag CONV_OP>
+using CKTransposeInputOp = TransposeOperand<ND, TranposeKind::NCHW_TO_NHWC, CONV_OP>;
+
+template <int ND, ConvOperandTag CONV_OP>
+using CKTransposeOutputOp = TransposeOperand<ND, TranposeKind::NHWC_TO_NCHW, CONV_OP>;
+
+class TransposeInstance
+{
+    size_t tensor_sz = 0;
+    std::vector<OpKernelArg> kern_args{};
+    size_t kern_idx   = std::numeric_limits<size_t>::max();
+    size_t buf_offset = 0;
+    shared<Data_t> buf_handle{};
+
+public:
+    template <typename TransSolnType>
+    TransposeInstance(const TransSolnType& trans_sol,
+                      size_t k_idx,
+                      const MultiBufferWorkspaceTraits& wt,
+                      size_t wspace_index)
+        : tensor_sz(trans_sol.GetOutputTensorSize()),
+          kern_args(trans_sol.GetKernelArg()),
+          kern_idx(k_idx),
+          buf_offset(wt.GetOffset(wspace_index))
+    {
+    }
+
+    void AssignBuffer(const Handle& handle, Data_t workSpace)
+    {
+        buf_handle = handle.CreateSubBuffer(workSpace, buf_offset, tensor_sz);
+        assert(buf_handle.get());
+    }
+
+    Data_t GetBufferPtr() const { return buf_handle.get(); }
+
+    void ConvertFrom(const Handle& handle, const std::vector<Kernel>& kernels, ConstData_t in_ptr)
+    {
+        Run(handle, kernels, buf_handle.get(), in_ptr);
+    }
+
+    void ConvertTo(const Handle& handle, const std::vector<Kernel>& kernels, Data_t out_ptr)
+    {
+        Run(handle, kernels, out_ptr, buf_handle.get());
+    }
+
+    void ZeroOutBuffer()
+    {
+        [[maybe_unused]] auto status = hipMemset(buf_handle.get(), 0, tensor_sz);
+        assert(status == hipSuccess);
+    }
+
+    TransposeInstance()                         = delete;
+    TransposeInstance(const TransposeInstance&) = default;
+    TransposeInstance(TransposeInstance&&)      = default;
+    ~TransposeInstance()                        = default;
+
+private:
+    void Run(const Handle& handle,
+             const std::vector<Kernel>& kernels,
+             Data_t out_ptr,
+             ConstData_t in_ptr)
+    {
+        assert(out_ptr);
+        assert(in_ptr);
+        assert(kernels.size() > kern_idx);
+
+        kern_args[0] = out_ptr;
+        kern_args[1] = in_ptr;
+
+        handle.Run(kernels[kern_idx])(kern_args);
+        if(handle.IsProfilingEnabled())
+        {
+            handle.AccumKernelTime(handle.GetKernelTime());
+        }
```

Comment:
You need to save previously accumulated time (because `handle.Run()` overwrites it):
```suggestion
        auto save = handle.IsProfilingEnabled() ? 0.0f : handle.GetKernelTime();
        handle.Run(kernels[kern_idx])(kern_args);
        if(handle.IsProfilingEnabled())
        {
            handle.AccumKernelTime(save);
        }
```

Reply:
So, this has been quite confusing. Partially because I wasn't paying attention to how kernel time was being handled via `Run()`. My question is:
Can we do better and avoid this confusion altogether? For example, why not have `Run()` automatically accumulate kernel times, so an invoker looks like: 
```
handle.ResetKernelTime();
handle.Run();
handle.Run();
...
```
Accumulation happens automatically via `elapsed_time_handler()`  here: https://github.com/ROCm/MIOpen/blob/5c7617c9241272cec6b6ae64fab18a59182f1b46/src/hip/handlehip.cpp#L206 , where all we need to change is update to accumulate.

Reply:
@amberhassaan 
> So, this has been quite confusing.

+1

> Can we do better and avoid this confusion altogether?

Sure. I've wanted to do this since the first day I saw this ugly API. The one only need free time and will to redesign this stuff and perform a lot of testing.

Reply:
I'll be motivated some day. A frustrated user is a motivated user :).

Reply:
[offtopic] @amberhassaan Haha, exactly! Wait... so that means I'm very motivated? ;))

Reply:
[resolved]

Reply:
#2736 created.